text/0000-command-ergonomics.md

```diff
@@ -0,0 +1,542 @@
+    fn stderr_bytes(&self) -> Option<&[u8]>;
+
+    /// If trouble included failure to spawn, the spawn error.
+    fn spawn_error() -> Option<&io::Error>;
+
+    /// If trouble included failure to talk to the child, the IO error.
+    ///
+    /// This might include problems which might be caused by child
+    /// misbehaviour.
+    fn communication_error() -> Option<&io::Error>;
+
+    /// If trouble included failed UTF-8 conversion.
+    fn utf8_error(&self) -> Option<&std::str::FromUtf8Error>;
+}
+```
+
+The `Display` implementation will print everything above
+including the command's arguments.
+The arguments will be escaped or quoted in some way that renders 
+a resulting error message unambiguous.
+
+### `impl From<SubprocessError> for io::Error`
+
+`SubprocessError` must be convertible to `io::Error`
+so that we can use it in `ChildOutputStream`'s
+`Read` implementation.
+This may also be convenient elsewhere.
+
+The `io::ErrorKind` for a `SubprocessError` will be:
+
+  * The `io::ErrorKind` from the spawn error, if any.
+
+  * Otherwise, a new kind `io::ErrorKind::ProcessFailed`,
+    which means that the subprocess itself failed.
+
+### Further necessary APIs for `SubprocessError`
+
+We also provide ways for this new error to be constructed,
+which will be needed by other lower level libraries besides std,
+notably async frameworks:
+
+```rust
+impl SubprocessError {
+    /// Makes a "blank" error which doesn't contain any useful information
+    ///
+    /// `has_problem()` will return `false` until one of the setters
+    /// is used to store an actual problem.
+    fn new_empty() -> Self { }
+
+    // If we keep ExitStatusError
+    fn from_exit_status_error(status: ExitStatusError) -> Self { }
+
+    fn set_program(&mut self, impl Into<OsString>);
+    fn set_args(&mut self, impl IntoIterator<Item=impl Into<OsString>>);
+    fn set_stdout_bytes(output: Option<Into<Box<[u8]>>>);
+
+    fn set_status(&mut self, status: ExitStatus);
+    fn set_stderr_bytes(&mut self, stderr: impl Into<Box<u8>>);
+    fn set_spawn_error(&mut self, error: Option<io::Error>);
+    fn set_communication_error(&mut self, error: Option<io::Error>);
+    fn set_utf8_error(&mut self, error: Option<std::str::FromUtf8Error>);
+
+    /// Find out if this error contains any actual error information
+    ///
+    /// Returns `false` for a fresh blank error,
+    /// or `true` for one which has any of the error fields set.
+    /// (Currently equivalent to checking all of `status()`,
+    /// `stderr_bytes()`, `spawn_error()` and `utf8_error()`.)
+    //
+    // We must provide this because it's needed for handling programs
+    // with unusual exit status conventions (eg `diff(1)`)
+    // and a caller can't reimplement it without making assumptons
+    // about `SubprocessError`'s contents.
+    fn has_problem(&self) -> bool;
+}
+impl Default for SubprocessError { ... }
+impl Clone for SubprocessError { ... } // contained io:Errors are in Arcs
```

Comment:
I think we should instead ask users who need a `Clone`able `SubprocessError` to use `Arc<SubprocessError>`, rather than imposing that requirement on the internals.

Reply:
Fair enough, I have changed this.